### DISCUSSION TITLE
Setting Axios Header to Java Backend based on value stored in express session

### DISCUSSION THREAD

**Discussion by NikhilNanjappa** [axios/axios#5955] [Q&A]
I have a node.js express (server-side rendering) frontend app that calls java backend endpoints. I use Axios to call the java endpoints. I need to set a header named "**agent-id**" into every request express makes to java, the value of that header comes from the express session (`req.session`).

My Axios code is on another module. I'm importing that module and calling my APIs.

Currently my Axios module code is setup like so:

```
const axiosClient = axios.create({
 baseURL: '...',
 httpsAgent: null,
 proxy: false,
 timeout: 60000,
});

axiosClient.interceptors.request.use(request => {
 console.log(`Starting:${request.baseURL}${request.url}`);

 // HOW DO I SET HEADER FROM EXPRESS SESSION HERE //

 return request;
}, error => Promise.reject(error));

axiosClient.interceptors.response.use(response => {
 console.log(`Received from:${response.config.baseURL}${response.config.url}`);
 return response;
}, error => Promise.reject(error));

module.exports = axiosClient;
```

and I call my APIs from frontend express app like so:

```
import api from 'my-api-module';

const { data = {} } = await api.someName();
```

**What I've tried**

I tried updating req.headers using a app.use middleware but I could not find that header within the axios interceptor.

Ideally I dont want to change my entire API calling structure by passing req obj on every existing API calls, so is there anyway I can globally set the header using express session value for every Java request ?

**Comment by HerrBastiii** [comment]
Hi, it's been almost a year, but how about setting the headers on axios client creation on your first line?

**Comment by jasonsaayman** [comment]
`req.headers` and Axios request headers are separate things.

`req.headers` are the headers from the incoming browser/request to your Express app. Updating them in Express middleware will not automatically affect the outgoing Axios request from Express to your Java backend.

Also, do **not** set this on a shared Axios singleton with `axiosClient.defaults.headers.common['agent-id'] = ...` per request. In SSR, multiple users can be handled at the same time, so a global/default header can leak one user's `agent-id` into another user's request.

You need a request-scoped value.

The cleanest option is to pass the value when making the request:

```js
await axiosClient.get('/some-api', {
  headers: {
    'agent-id': req.session.agentId
  }
});
```

If you do not want to pass `req` or `agentId` through every API function, use Node's `AsyncLocalStorage` to keep request context available to your Axios interceptor.

Example:

```js
// requestContext.js
const { AsyncLocalStorage } = require('node:async_hooks');

const requestContext = new AsyncLocalStorage();

module.exports = { requestContext };
```

```js
// express app
const { requestContext } = require('./requestContext');

app.use((req, res, next) => {
  requestContext.run(
    {
      agentId: req.session?.agentId
    },
    next
  );
});
```

Then in your Axios module:

```js
const axios = require('axios');
const { requestContext } = require('./requestContext');

const axiosClient = axios.create({
  baseURL: '...',
  httpsAgent: null,
  proxy: false,
  timeout: 60000
});

axiosClient.interceptors.request.use(
  config => {
    const agentId = requestContext.getStore()?.agentId;

    if (agentId) {
      if (config.headers?.set) {
        config.headers.set('agent-id', agentId);
      } else {
        config.headers = {
          ...config.headers,
          'agent-id': agentId
        };
      }
    }

    return config;
  },
  error => Promise.reject(error)
);

module.exports = axiosClient;
```

Now your existing API calls can stay roughly the same:

```js
import api from 'my-api-module';

const { data = {} } = await api.someName();
```

As long as `api.someName()` is called during the same Express request async chain, the interceptor can read the current request's `agentId`.

So the short answer is: Axios cannot magically read `req.session` from a separate module. Either pass the value explicitly, create a per-request Axios/API client, or use `AsyncLocalStorage` for request-scoped context. Avoid global Axios header mutation for session-specific values.
